src/ucs/datastruct/ptr_map.inl

```diff
@@ -33,134 +33,174 @@ BEGIN_C_DECLS
-           UCS_PTR_MAP_KEY_INDIRECT_FLAG;
-
-    iter = kh_put(ucs_ptr_map_impl, &map->hash, *key, &ret);
+    iter = kh_put(ucs_ptr_map_impl, hash, key, &ret);
     if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
         return UCS_ERR_NO_MEMORY;
     } else if (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT)) {
         return UCS_ERR_ALREADY_EXISTS;
     }
 
-    kh_value(&map->hash, iter) = ptr;
+    kh_value(hash, iter) = ptr;
     return UCS_OK;
 }
 
-/**
- * Get a pointer value from the map by its key.
- *
- * @param [in]  map     Container to get the pointer value from.
- * @param [in]  key     Key to look up in the container.
- * @param [in]  extract Whether to remove the key from the map.
- * @param [out] ptr_p   If successful, set to the pointer found in the map.
- *
- * @return UCS_OK if found, UCS_ERR_NO_PROGRESS if this key is direct and
- *         therefore no action was performed, UCS_ERR_NO_ELEM if not found.
- */
-static UCS_F_ALWAYS_INLINE ucs_status_t
-ucs_ptr_map_get(ucs_ptr_map_t *map, ucs_ptr_map_key_t key, int extract,
-                void **ptr_p)
+static inline ucs_status_t
+ucs_ptr_hash_get(ucs_ptr_hash_t *hash, ucs_ptr_map_key_t key, int extract,
+                 void **ptr_p)
 {
     khiter_t iter;
 
-    if (ucs_likely(!ucs_ptr_map_key_indirect(key))) {
-        *ptr_p = (void*)key;
-        return UCS_ERR_NO_PROGRESS;
-    }
-
-    iter = kh_get(ucs_ptr_map_impl, &map->hash, key);
-    if (ucs_unlikely(iter == kh_end(&map->hash))) {
+    iter = kh_get(ucs_ptr_map_impl, hash, key);
+    if (ucs_unlikely(iter == kh_end(hash))) {
         return UCS_ERR_NO_ELEM;
     }
 
-    *ptr_p = kh_value(&map->hash, iter);
+    *ptr_p = kh_value(hash, iter);
     if (extract) {
-        kh_del(ucs_ptr_map_impl, &map->hash, iter);
+        kh_del(ucs_ptr_map_impl, hash, iter);
     }
     return UCS_OK;
 }
 
-/**
- * Remove object pointer from the map by key.
- *
- * @param [in]  map     Container.
- * @param [in]  key     Key to object pointer.
- *
- * @return - UCS_OK on success
- *         - UCS_ERR_NO_PROGRESS if this key is direct and therefore no action
- *           was performed
- *         - UCS_ERR_NO_ELEM if the key is not found in the internal hash table.
- */
-static UCS_F_ALWAYS_INLINE ucs_status_t
-ucs_ptr_map_del(ucs_ptr_map_t *map, ucs_ptr_map_key_t key)
+static inline int
+ucs_ptr_check_direct_put(void *ptr, int direct, ucs_ptr_map_key_t *key)
+{
+    if (ucs_unlikely(!direct)) {
+        return 0;
+    }
+
+    *key = (uintptr_t)ptr;
+    ucs_assert(!(*key & UCS_PTR_MAP_KEY_MIN_ALIGN));
+    ucs_assert(*key != 0);
+    return 1;
+}
+
+static inline int
+ucs_ptr_check_direct_get(ucs_ptr_map_key_t key, void **ptr_p)
 {
-    void UCS_V_UNUSED *dummy;
-    return ucs_ptr_map_get(map, key, 1, &dummy);
+    if (ucs_unlikely(ucs_ptr_map_key_indirect(key))) {
+        return 0;
+    }
+
+    *ptr_p = (void*)key;
+    return 1;
 }
 
+#define UCS_PTR_MAP_PUT_IMPL_0(_map, _ptr, _indirect, _key) \
```

Comment:
could be done w/o macro: pass safe/unsafe as input param to all functions. the compiler should be able to resolve the constants.

Reply:
Sorry, I do not get your point.
The macros allows to avoid incorrect combination of put, get, del methods. A user defines which implementation to use once using the macro parameter instead of using function parameter for each function call.

Reply:
IMO it's better that more code will be in regular inline functions rather than in macro body. it helps with debugging since line numbers are not mixed up.
for example, maybe can do something like "if (safe) { safe_map = container_of(unsafe_map);  [do-safe-stuff] } in put/get operations, where "safe" is a boolean parameter of put/get/etc.
and the named PUT/GET operations will just call the inline put/get with appropriate parameters

Reply:
👍 for inline where it's possible, it also could reduce some code duplication

Reply:
Updated implementation of methods working with ptr_map. Major part of code is in inline functions.